### PR TITLE
Adding support for std::array

### PR DESCRIPTION
**Pull request by henryiii**
This should add support for std::array, which is missing (though inferior old C style arrays were supported!).

I tried to add a test, as well.

Edit: Addresses #100.

### CODE REVIEW

**Comment by henryiii** [comment]
Looks like I got unlucky and copied a test that wasn't run (missing BOOST macro). Will try to fix.

**Comment by henryiii** [comment]
Fixed tests and code. Currently only supports 1D (non-nested) `std::array`s. This could be expanded similar to `std::vector`, but for now this is simpler and fits my use case. The test for vectors has been restored, as well.

File: include/highfive/bits/H5Utils.hpp [hunk 1]
```diff
@@ -57,6 +58,12 @@ struct array_dims<T[N]> {
     static const size_t value = 1 + array_dims<T>::value;
 };
 
+// Only supporting 1D arrays at the moment
+template<typename T, std::size_t N>
+struct array_dims<std::array<T,N>> {
+    static const size_t value = 1;
```
Comment (hernando) [inline include/highfive/bits/H5Utils.hpp]: Shoulnd't it be?

    static const size_t value = 1 + array_dims<T>::value;

Comment (hernando) [reply include/highfive/bits/H5Utils.hpp]: All the const can be constexpr

File: include/highfive/bits/H5Converter_misc.hpp [hunk 2]
```diff
@@ -174,6 +175,34 @@ struct data_converter<
...
+    
+    inline typename type_of_array<T>::type*
+    transform_write(std::array<T, S>& vec) {
+        return vec.data();
+    }
+    
+    inline void process_result(std::array<T, S>& vec) { (void)vec; }
+    
+    DataSpace* _space;
+    size_t _dim;
```
Comment (hernando) [inline include/highfive/bits/H5Converter_misc.hpp]: Members not needed, remove.

File: include/highfive/bits/H5Converter_misc.hpp [hunk 3]
```diff
@@ -174,6 +175,34 @@ struct data_converter<
...
+// apply conversion to std::array
+template <typename T, std::size_t S>
+struct data_converter<
+    std::array<T, S>,
+    typename std::enable_if<(
+                         std::is_same<T, typename type_of_array<T>::type>::value)>::type> {
+    inline data_converter(std::array<T,S>& vec, DataSpace& space, size_t dim = 0)
+    : _space(&space), _dim(dim) {
+        assert(_space->getDimensions().size() >= dim);
+        (void)vec;
```
Comment (hernando) [inline include/highfive/bits/H5Converter_misc.hpp]: All other types are resizing the back storage according to `space` except in the case of plain C arrays, where not even the size can be checked. Since here you can't resize, but at least you can check the size, you should verify that space.getDimensions()[0] == S and throw DataSpaceException otherwise. Exercise this in the test.

File: include/highfive/bits/H5Utils.hpp (hunk 1 above)
Comment (henryiii) [reply include/highfive/bits/H5Utils.hpp]: Change the existing ones too? `static constexpr size_t`?

File: include/highfive/bits/H5Converter_misc.hpp [hunk 4]
```diff
@@ -174,6 +175,34 @@ struct data_converter<
...
+    }
+    
+    inline typename type_of_array<T>::type*
+    transform_read(std::array<T, S>& vec) {
+        return &(vec[0]);
```
Comment (hernando) [inline include/highfive/bits/H5Converter_misc.hpp]: Why not vec.data()?

File: include/highfive/bits/H5Utils.hpp (hunk 1 above)
Comment (hernando) [reply include/highfive/bits/H5Utils.hpp]: Yes, I'd do it

File: include/highfive/bits/H5Converter_misc.hpp (hunk 4 above)
Comment (henryiii) [reply include/highfive/bits/H5Converter_misc.hpp]: That's how `std::vector` was done. I changed one of them to `.data`, will do the other too.

File: include/highfive/bits/H5Utils.hpp (hunk 1 above)
Comment (henryiii) [reply include/highfive/bits/H5Utils.hpp]: Done.

File: tests/unit/tests_high_five_base.cpp [hunk 5]
```diff
@@ -647,10 +647,12 @@ void readWriteVectorTest() {
     std::ostringstream filename;
     filename << "h5_rw_vec_" << typeid(T).name() << "_test.h5";
 
-    std::srand((unsigned int)std::time(0));
     const size_t x_size = 800;
     const std::string DATASET_NAME("dset");
     typename std::vector<T> vec;
+    vec.resize(x_size);
+    for(size_t i=0; i<x_size; i++)
```
[the remark this answers is not in the document]
Comment (henryiii) [reply tests/unit/tests_high_five_base.cpp]: Assuming this is about the spacing, done. Also changed the spacing in the array test to match.

File: include/highfive/bits/H5Converter_misc.hpp (hunk 2 above)
Comment (henryiii) [reply include/highfive/bits/H5Converter_misc.hpp]: Dropped the `_dim`. I think `_space` is still needed?

File: include/highfive/bits/H5Converter_misc.hpp (hunk 3 above)
Comment (henryiii) [reply include/highfive/bits/H5Converter_misc.hpp]: Changed, see if this is better.

File: include/highfive/bits/H5Converter_misc.hpp (hunk 4 above)
Comment (henryiii) [reply include/highfive/bits/H5Converter_misc.hpp]: Done (only changed the arrays, not the vectors).

File: include/highfive/bits/H5Converter_misc.hpp (hunk 3 above)
Comment (hernando) [reply include/highfive/bits/H5Converter_misc.hpp]: You don't need to use 'at' if you have already checked that the size is not 0. But what I was suggesting is to have an if and throw an exception. In H5Slice_traits_misc.hpp you can see that the dimensions are checked and an exception is if they don't match. I don't see a good reason to not doing the same thing here as the penalty of checking that compared to the IO operation is going to be negligible. I've realized that the assert is not needed because of the check made by the caller, but I know the other converters are doing it.

File: tests/unit/tests_high_five_base.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hernando) [reply tests/unit/tests_high_five_base.cpp]: Yes, it was

File: include/highfive/bits/H5Converter_misc.hpp (hunk 2 above)
Comment (hernando) [reply include/highfive/bits/H5Converter_misc.hpp]: In the other (when it's used) it appears in process_results, transform_read or transform_write. See the C array converter, it doesn't have anything.

Comment (henryiii) [reply include/highfive/bits/H5Converter_misc.hpp]: Removed.

File: include/highfive/bits/H5Converter_misc.hpp (hunk 3 above)
Comment (henryiii) [reply include/highfive/bits/H5Converter_misc.hpp]: Is the new version better?

**Comment by henryiii** [comment]
Great, thanks! If anyone is interested, a good next step would be to extend this to nested arrays. (For me, I'm using `multi_array` from Boost instead of nested arrays, so this is sufficient).

File: tests/unit/tests_high_five_base.cpp [hunk 6]
```diff
@@ -667,11 +669,48 @@ void readWriteVectorTest() {
...
+    File file(filename.str(), File::ReadWrite | File::Create | File::Truncate);
+
+    // Create a dataset with double precision floating points
+    DataSet dataset = file.createDataSet<T>(DATASET_NAME, DataSpace::From(vec));
+
+    dataset.write(vec);
+
+    typename std::array<T, x_size> result;
+
+    dataset.read(result);
```
Comment (hernando) [inline tests/unit/tests_high_five_base.cpp]: Here I forgot to ask for

    typename std::array<T, 1> tooSmall;
    BOOST_CHECK_THROW(dataset.read(tooSmall), DataSpaceException)

Do you mind creating another PR for it?
